### PR TITLE
Basesvd optimization

### PR DESCRIPTION
**Pull request by eozd**
This PR implements loop unrolling, scalar replacement and vectorization for base SVD code. Two new functions are of particular importance:
1. `svd_subprocedure_vectorized`: Implement vectorizations by doing the updates in the same way as in base algorithm.
2. `svd_subprocedure_vectorized_transpose`: Take the transpose in subprocedure itself before doing the updates. This changes 3 column and 1 row updates to 3 row and 1 column updates which presumably should be more efficient.

Since AVX2 vectorizations require 32 byte aligned memory, we use `boost::align::aligned_allocator` as the allocator object of vectors. Additionally, `aligned_alloc` is used in place of `malloc` calls. We rely on [cmake-boost](https://github.com/Orphis/boost-cmake) to download and use boost automatically. I had to modify boost-cmake a little to make it work with header-only parts.

### CODE REVIEW

**Comment by ktrianta** [comment]
What changes diid you observe regarding performance after adding these optimizations?

**Comment by eozd** [comment]
I haven't done any extensive benchmarks. We can start measuring and tweaking the codes after these and the matrix multiplication codes are merged together

**Comment by ktrianta** [comment]
> I haven't done any extensive benchmarks. We can start measuring and tweaking the codes after these and the matrix multiplication codes are merged together

I did some quick benchmarking on the optimized base svd and did not observe an increase in performance as big as I would expect. I will try to do some more benchmarks today so we can check them tomorrow. If the results are not what we expect maybe we could look a bit into the Eigen code, as they achieve somewhat higher performance from our original base without using blocking.

**Comment by eozd** [comment]
Did you use it in blocked version? If so, it wouldn't have an effect without matrix multiplications.

**Comment by eozd** [comment]
Btw, my main optimized base svd is in transposed pull request. I hope that code together with matrix multiplications will outperform these versions because we operator mostly on rows.

**Comment by ktrianta** [comment]
> Did you use it in blocked version? If so, it wouldn't have an effect without matrix multiplications.

I had missed something so the results were off. Even in the blocking version without optimized multiplication there is improvement in performance. Why do you think it wouldn't have an effect without matrix multiplications?

**Comment by eozd** [comment]
Subprocedure works only on `2b x 2b` matrices while we do multiplications over the whole big matrix. So the main bottleneck should be the multiplications I think.

File: src/svd/two-sided/svd_subprocedure_vectorized.cpp [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+                const double c1k = cf.c1 * cf.k;
+
+                size_t k = 0;
+                __m256d vi_0, vi_1, vj_0, vj_1;
+                __m256d left_0, left_1, right_0, right_1;
+                __m256d v_c = _mm256_set1_pd(cf.c1);
+                __m256d v_s = _mm256_set1_pd(cf.s1);
+                __m256d v_s1k = _mm256_set1_pd(s1k);
+                __m256d v_c1k = _mm256_set1_pd(c1k);
+                for (; k + 7 < n; k += 8) {
```
Comment (RajaSimha) [inline src/svd/two-sided/svd_subprocedure_vectorized.cpp]: We might need to check different unroll counts!

**Pull request merged**